pkg/util/resourceexecutor/updater.go

```diff
@@ -0,0 +1,247 @@
+}
+
+func (u *DefaultResourceUpdater) ResourceType() sysutil.ResourceType {
+	return sysutil.ResourceType(u.file)
+}
+
+func (u *DefaultResourceUpdater) Path() string {
+	return filepath.Join(u.dir, u.file) // no additional parent dir here
+}
+
+func (u *DefaultResourceUpdater) Value() string {
+	return u.value
+}
+
+func (u *DefaultResourceUpdater) Update() error {
+	return u.updateFunc(u)
+}
+
+func (u *DefaultResourceUpdater) MergeUpdate() (ResourceUpdater, error) {
+	return nil, u.updateFunc(u)
+}
+
+func (u *DefaultResourceUpdater) GetLastUpdateTimestamp() time.Time {
+	return u.lastUpdateTimestamp
+}
+
+func (u *DefaultResourceUpdater) UpdateLastUpdateTimestamp(time time.Time) {
+	u.lastUpdateTimestamp = time
+}
+
+func CommonDefaultUpdateFunc(resource ResourceUpdater) error {
+	c := resource.(*DefaultResourceUpdater)
+	_ = audit.V(5).Reason(ReasonUpdateSystemConfig).Message("update %v to %v", resource.Path(), resource.Value()).Do()
+	return sysutil.CommonFileWriteIfDifferent(c.Path(), c.value)
+}
+
+// NewCommonDefaultUpdater returns a DefaultResourceUpdater for update general files.
+func NewCommonDefaultUpdater(file string, dir string, value string) (ResourceUpdater, error) {
+	return &DefaultResourceUpdater{
+		file:       file,
+		dir:        dir,
+		value:      value,
+		updateFunc: CommonDefaultUpdateFunc,
+	}, nil
+}
+
+type GuestCgroupResourceUpdater struct {
+	*CgroupResourceUpdater
+	sandboxID string // for execute the file operation inside the sandbox
+}
+
+type NewResourceUpdaterFunc func(resourceType sysutil.ResourceType, parentDir string, value string) (ResourceUpdater, error)
+
+type ResourceUpdaterFactory interface {
+	Register(g NewResourceUpdaterFunc, resourceTypes ...sysutil.ResourceType)
+	New(resourceType sysutil.ResourceType, parentDir string, value string) (ResourceUpdater, error)
+}
+
+func CommonCgroupUpdateFunc(resource ResourceUpdater) error {
+	c := resource.(*CgroupResourceUpdater)
+	// TODO: add audit
+	_ = audit.V(5).Reason(ReasonUpdateCgroups).Message("update %v to %v", resource.Path(), resource.Value()).Do()
+	return sysutil.CgroupFileWriteIfDifferent(c.parentDir, c.file, c.value)
+}
+
+// NewCommonCgroupUpdater returns a CgroupResourceUpdater for updating known cgroup resources.
+func NewCommonCgroupUpdater(resourceType sysutil.ResourceType, parentDir string, value string) (ResourceUpdater, error) {
+	r, ok := sysutil.DefaultRegistry.Get(sysutil.GetCurrentCgroupVersion(), resourceType)
+	if !ok {
+		return nil, fmt.Errorf("%s not found in cgroup registry", resourceType)
+	}
+	return &CgroupResourceUpdater{
+		file:       r,
+		parentDir:  parentDir,
+		value:      value,
+		updateFunc: CommonCgroupUpdateFunc,
+	}, nil
+}
+
+type CgroupUpdaterFactoryImpl struct {
+	lock     sync.RWMutex
+	registry map[sysutil.ResourceType]NewResourceUpdaterFunc
+}
+
+func NewCgroupUpdaterFactory() ResourceUpdaterFactory {
+	return &CgroupUpdaterFactoryImpl{
+		registry: map[sysutil.ResourceType]NewResourceUpdaterFunc{},
```

Comment:
how to separate cgroup v1/v2 of resource updater.

Reply:
The CgroupResource is registered in system.DefaultRegistry with correct CgroupVersion. Here the DefaultCgroupUpdaterFactory retrieves the CgroupResource to make up Updater.